### ISSUE TITLE
idea: optionally have generated functions be methods on a type

### ISSUE THREAD

**Issue by andre1sk** [a-h/templ#174]
To make theming easy given a flag templ could generate methods on a type
So one can define an interface for a theme

**Comment by a-h** [comment]
I'm not 100% sure what you mean, but I think this is already possible?

```go
package testmethod

type Data struct {
	message string
}

templ (d Data) Method() {
	<div>{ d.message }</div>
}
```

```go
package testmethod

import (
	_ "embed"
	"testing"

	"github.com/a-h/templ/generator/htmldiff"
)

//go:embed expected.html
var expected string

func Test(t *testing.T) {
	d := Data{
		message: "You can implement methods on a type.",
	}
	component := d.Method()

	diff, err := htmldiff.Diff(component, expected)
	if err != nil {
		t.Fatal(err)
	}
	if diff != "" {
		t.Error(diff)
	}
}
```

Would like to understand your use case though!

**Comment by andre1sk** [comment]
@a-h  oh I am sorry did not realize this was supported. TEMPL is amazing btw. enables productivity level I've never experienced in Go projects before. The use case is having themable frontend where people can implement an interface to create a compatible theme

**Comment by a-h** [comment]
No need to apologise. The docs definitely don't mention it. Maybe could update it!

Thanks for your comments on templ, glad you're finding it useful!

**Comment by a-h** [comment]
The action here is to update the docs in the relevant place to explain that it's possible to add methods that return templ Components to existing types.

**Comment by byt3h3ad** [comment]
hey! i wanna work on this!

**Comment by a-h** [comment]
Closed in #267
